File: LittleFencer-Android/app/src/main/java/com/littlefencer/app/logic/ActionDetectorManager.kt
package com.littlefencer.app.logic

import com.littlefencer.app.logic.detectors.*

/**
 * ActionDetectorManager - Orchestrates all action detectors.
 * 
 * Manages a pool of specialized detectors, processes each frame through
 * all relevant detectors, and returns the most confident action.
 * 
 * Priority handling:
 * - Completed actions take priority over in-progress
 * - Higher confidence in-progress actions take priority
 * - Conflicting detections resolved by action hierarchy
 */
class ActionDetectorManager {

    private val detectors: List<ActionDetector> = listOf(
        AdvanceDetector(),
        RetreatDetector(),
        LungeDetector(),          // NEW: Independent lunge detector
        AdvanceLungeDetector(),
        BalestraLungeDetector(),
        FlungeDetector(),
        ParryRiposteDetector()
    )

    // Pose history buffer
    private val poseHistory = mutableListOf<PoseFrame>()
    private val maxHistorySize = 30  // ~1 second at 30fps

    /**
     * Process a new frame through all detectors.
     * 
     * @param landmarks Current frame landmarks
     * @param timestampMs Frame timestamp
     * @return Best action result from all detectors
     */
    fun processFrame(landmarks: List<android.graphics.PointF>, timestampMs: Long): ActionResult {
        val currentFrame = PoseFrame(landmarks, timestampMs)
        
        // Add to history
        poseHistory.add(currentFrame)
        while (poseHistory.size > maxHistorySize) {
            poseHistory.removeAt(0)
        }
        
        // Need minimum history
        if (poseHistory.size < 5) {
            return ActionResult.None
        }
        
        // Run all detectors
        val results = detectors.map { detector ->
            detector.detect(currentFrame, poseHistory.dropLast(1))
        }
        
        // Find best result (Completed > InProgress > None)
        val completedResults = results.filterIsInstance<ActionResult.Completed>()
        if (completedResults.isNotEmpty()) {
            // Return highest quality completed action
            return completedResults.maxByOrNull { 
                when (it.quality) {
                    ActionQuality.PERFECT -> 4
                    ActionQuality.GOOD -> 3
                    ActionQuality.ACCEPTABLE -> 2
                    ActionQuality.POOR -> 1
                }
            } ?: completedResults.first()
        }
        
        val inProgressResults = results.filterIsInstance<ActionResult.InProgress>()
        if (inProgressResults.isNotEmpty()) {
            // Return highest confidence in-progress action
            return inProgressResults.maxByOrNull { it.confidence } ?: inProgressResults.first()
        }
        
        return ActionResult.None
    }

    /**
     * Get information about all currently in-progress actions.
     * Useful for UI to show multiple potential actions.
     */
    fun getInProgressActions(): List<ActionResult.InProgress> {
        val currentFrame = poseHistory.lastOrNull() ?: return emptyList()
        val history = poseHistory.dropLast(1)
        
        if (history.size < 5) return emptyList()
        
        return detectors.mapNotNull { detector ->
            val result = detector.detect(currentFrame, history)
            result as? ActionResult.InProgress
        }
    }

    /**
     * Reset all detectors (e.g., when user leaves frame).
     */
    fun reset() {
        detectors.forEach { it.reset() }
        poseHistory.clear()
    }

    /**
     * Get list of all supported actions.
     */
    fun getSupportedActions(): List<SaberAction> = detectors.map { it.targetAction }
}

File: LittleFencer-iOS/LittleFencer/Logic/ActionDetectorManager.swift
import CoreGraphics
import Foundation

/// Orchestrates all action detectors
///
/// Manages a pool of specialized detectors, processes each frame through
/// all relevant detectors, and returns the most confident action.
///
/// Priority handling:
/// - Completed actions take priority over in-progress
/// - Higher confidence in-progress actions take priority
/// - Conflicting detections resolved by action hierarchy
class ActionDetectorManager {
    
    // MARK: - Detectors
    
    private let detectors: [ActionDetector] = [
        AdvanceDetector(),
        LungeDetector(),          // Independent lunge detector
        RetreatDetector(),
        AdvanceLungeDetector(),
        BalestraLungeDetector(),
        FlungeDetector(),
        ParryRiposteDetector()
    ]
    
    // MARK: - Pose History
    
    private var poseHistory: [PoseFrame] = []
    private let maxHistorySize = 30  // ~1 second at 30fps
    
    // MARK: - Processing
    
    /// Process a new frame through all detectors
    func processFrame(landmarks: [CGPoint], timestampMs: Double) -> ActionResult {
        let currentFrame = PoseFrame(landmarks: landmarks, timestampMs: timestampMs)
        
        // Add to history
        poseHistory.append(currentFrame)
        while poseHistory.count > maxHistorySize {
            poseHistory.removeFirst()
        }
        
        // Need minimum history
        guard poseHistory.count >= 5 else {
            return .none
        }
        
        // Run all detectors
        let results = detectors.map { detector in
            detector.detect(current: currentFrame, history: Array(poseHistory.dropLast()))
        }
        
        // Find best result (Completed > InProgress > None)
        var completedResults: [(ActionResult, ActionQuality)] = []
        for result in results {
            if case let .completed(_, quality, _, _) = result {
                completedResults.append((result, quality))
            }
        }
        
        if !completedResults.isEmpty {
            // Return highest quality completed action
            if let best = completedResults.max(by: { $0.1 < $1.1 }) {
                return best.0
            }
            return completedResults[0].0
        }
        
        // Check for in-progress results
        var inProgressResults: [(ActionResult, CGFloat)] = []
        for result in results {
            if case let .inProgress(_, confidence, _) = result {
                inProgressResults.append((result, confidence))
            }
        }
        
        if !inProgressResults.isEmpty {
            // Return highest confidence in-progress action
            if let best = inProgressResults.max(by: { $0.1 < $1.1 }) {
                return best.0
            }
            return inProgressResults[0].0
        }
        
        return .none
    }
    
    /// Get information about all currently in-progress actions
    func getInProgressActions() -> [ActionResult] {
        guard let currentFrame = poseHistory.last,
              poseHistory.count >= 5 else {
            return []
        }
        
        let history = Array(poseHistory.dropLast())
        
        return detectors.compactMap { detector -> ActionResult? in
            let result = detector.detect(current: currentFrame, history: history)
            if case .inProgress = result {
                return result
            }
            return nil
        }
    }
    
    /// Reset all detectors
    func reset() {
        detectors.forEach { $0.reset() }
        poseHistory.removeAll()
    }
    
    /// Get list of all supported actions
    func getSupportedActions() -> [SaberAction] {
        return detectors.map { $0.targetAction }
    }
}
